File: rust/plugin-runtime/tests/package_source.cpp
// Copyright (c) 2026 OceanBase. Licensed under the Apache License, Version 2.0.
// Real C++ source adapter -> Rust parser/filesystem/owned snapshot. No SQL server.
#include "share/plugin/extension_package.h"
#include "lib/ob_errno.h"
#include <cstdlib>
#include <fstream>
#include <iostream>

#define CHECK(expr) do { if (!(expr)) { std::cerr << __LINE__ << ": " << #expr << std::endl; std::abort(); } } while (false)

int main(int argc, char **argv)
{
  CHECK(argc == 3);
  using namespace oceanbase::common;
  using namespace oceanbase::share::plugin;
  ExtensionPackageSource source;
  std::string error;
  // Real C++ layout -> Rust in-memory source, with no filesystem entry point.
  ExtensionPackageSource memory;
  memory.name_ = "memory_ops";
  memory.version_ = "2";
  memory.native_module_ = "org.memory";
  memory.requires_ = {"base"};
  memory.prerequisites_ = {"migration"};
  memory.scripts_ = {{"", "1", "SELECT 'a;b'; -- no newline"}, {"1", "2", ""}};
  CHECK(validate_extension_package_source(memory, error) == OB_SUCCESS && error.empty());
  for (int test = 0; test < 16; ++test) {
    auto bad = memory;
    switch (test) {
      case 0: bad.name_ = "../escape"; break;
      case 1: bad.version_ = "3"; break;
      case 2: bad.scripts_[1].from_version_ = "another"; break;
      case 3: bad.scripts_[0].sql_ = " \n"; break;
      case 4: bad.scripts_[0].sql_ = std::string(1, '\xff'); break;
      case 5: bad.scripts_[0].sql_ = std::string("x\0y", 3); break;
      case 6: bad.scripts_[0].sql_.assign(4 * 1024 * 1024, 'x'); bad.scripts_[1].sql_ = "x"; break;
      case 7: bad.requires_ = {"base", "base"}; break;
      case 8: bad.requires_ = {"memory_ops"}; break;
      case 9: bad.native_module_ = "Other.Module"; break;
      case 10: bad.schema_ = "fixed"; bad.relocatable_ = true; break;
      case 11: bad.from_version_ = "2"; break;
      case 12: bad.prerequisites_ = {"base"}; break;
      case 13: bad.prerequisites_ = {"memory_ops"}; break;
      case 14: bad.prerequisites_ = {"migration", "migration"}; break;
      case 15: bad.prerequisites_.assign(64, "migration"); break;
    }
    CHECK(validate_extension_package_source(bad, error) == OB_INVALID_ARGUMENT && !error.empty());
  }
  memory.from_version_ = "1";
  memory.scripts_.erase(memory.scripts_.begin());
  CHECK(validate_extension_package_source(memory, error) == OB_SUCCESS);
  memory.from_version_ = "2";
  memory.scripts_.clear();
  CHECK(validate_extension_package_source(memory, error) == OB_INVALID_ARGUMENT);
  memory.prerequisites_.clear();
  CHECK(validate_extension_package_source(memory, error) == OB_SUCCESS);
  memory.from_version_.clear();
  CHECK(validate_extension_package_source(memory, error) == OB_INVALID_ARGUMENT);
  const std::string root = argv[1];
  CHECK(read_extension_package(root, "native_only", "", source, error) == OB_SUCCESS);
  CHECK(source.native_install_ && source.native_module_ == "org.test" && source.version_ == "1.0" && source.scripts_.empty());
  CHECK(validate_extension_package_source(source, error) == OB_SUCCESS);
  auto native_source = source;
  native_source.native_install_ = false;
  CHECK(validate_extension_package_source(native_source, error) == OB_INVALID_ARGUMENT);
  native_source = source; native_source.native_module_.clear();
  CHECK(validate_extension_package_source(native_source, error) == OB_INVALID_ARGUMENT);
  CHECK(read_extension_package(root, "native_only", "1.1", source, error) == OB_FILE_NOT_EXIST && source.name_.empty());
  CHECK(read_extension_update(root, "native_only", "1.0", "1.1", source, error) == OB_SUCCESS);
  CHECK(source.native_install_ && source.scripts_.size() == 1 && source.scripts_[0].sql_ == "SELECT 2;");
  CHECK(validate_extension_package_source(source, error) == OB_SUCCESS);
  CHECK(read_extension_update(root, "native_only", "1.0", "1.0", source, error) == OB_SUCCESS);
  CHECK(source.native_install_ && source.scripts_.empty() && source.from_version_ == "1.0");
  CHECK(read_extension_package(root, "plain", "", source, error) == OB_SUCCESS);
  CHECK(source.name_ == "plain" && source.version_ == "1.0");
  CHECK(source.native_module_.empty() && source.schema_.empty() && source.relocatable_);
  CHECK(source.requires_ == std::vector<std::string>({"dep1", "dep2"}));
  CHECK(source.scripts_.size() == 1 && source.scripts_[0].sql_ == "SELECT 'first';\n");
  CHECK(source.scripts_[0].from_version_.empty() && source.scripts_[0].to_version_ == "1.0");
  // Returned strings outlive the Rust handle and later filesystem changes.
  std::ofstream(root + "/plain/plain--1.0.sql") << "SELECT 'changed';\n";
  CHECK(source.scripts_[0].sql_ == "SELECT 'first';\n");
  CHECK(read_extension_package(root, "plain", "2.0", source, error) == OB_SUCCESS);
  CHECK(source.version_ == "2.0" && source.scripts_[0].sql_ == "SELECT 'second';\n");
  for (const char *name : {"escape", "script_escape", "bad_utf8", "sql_nul", "empty_sql",
                          "control_utf8", "control_large", "large", "bad_control", "directory", "selfdep"}) {
    CHECK(read_extension_package(root, name, "", source, error) == OB_INVALID_ARGUMENT);
    CHECK(source.name_.empty() && source.scripts_.empty() && source.requires_.empty());
    CHECK(!error.empty());
  }
  CHECK(read_extension_package(root, "exact_limit", "", source, error) == OB_SUCCESS);
  CHECK(source.scripts_[0].sql_.size() == 4 * 1024 * 1024);
  CHECK(read_extension_package(std::string(32769, 'x'), "plain", "", source, error) == OB_INVALID_ARGUMENT);
  CHECK(source.scripts_.empty() && !error.empty());
  CHECK(read_extension_package(root, "../plain", "", source, error) == OB_INVALID_ARGUMENT);
  CHECK(read_extension_package(root, "plain", "../2.0", source, error) == OB_INVALID_ARGUMENT);
  CHECK(read_extension_package(root, "plain", "99", source, error) == OB_FILE_NOT_EXIST);
  CHECK(read_extension_package(root, "no_default", "", source, error) == OB_INVALID_ARGUMENT);
  CHECK(read_extension_package(root, "no_default", "1", source, error) == OB_SUCCESS);
  CHECK(read_extension_package(root, "chain", "", source, error) == OB_SUCCESS);
  CHECK(source.version_ == "tip" && source.scripts_.size() == 3);
  CHECK(source.scripts_[0].from_version_.empty() && source.scripts_[0].to_version_ == "1.0");
  CHECK(source.scripts_[0].sql_ == "SELECT 'base'; -- tail without newline");
  CHECK(source.scripts_[1].from_version_ == "1.0" && source.scripts_[1].to_version_ == "middle");
  CHECK(source.scripts_[1].sql_ == "SELECT 'middle';");
  CHECK(source.scripts_[2].from_version_ == "middle" && source.scripts_[2].to_version_ == "tip");
  CHECK(source.scripts_[2].sql_ == "SELECT 'tip';");
  std::ofstream(root + "/chain/chain--middle--tip.sql") << "SELECT 'changed';";
  CHECK(source.scripts_[2].sql_ == "SELECT 'tip';");
  CHECK(read_extension_package(root, "chain", "direct", source, error) == OB_SUCCESS);
  CHECK(source.scripts_.size() == 1 && source.scripts_[0].sql_ == "SELECT 'direct';");
  CHECK(read_extension_package(root, "chain_override", "", source, error) == OB_SUCCESS);
  CHECK(source.requires_ == std::vector<std::string>({"hidden_dependency"}) && source.prerequisites_.empty());
  for (const char *name : {"chain_escape", "chain_invalid", "chain_total", "secondary_default",
                          "secondary_directory", "secondary_self", "secondary_schema",
                          "secondary_module", "secondary_large", "secondary_escape", "dependency_overflow"}) {
    CHECK(read_extension_package(root, name, "", source, error) == OB_INVALID_ARGUMENT);
    CHECK(source.name_.empty() && source.scripts_.empty() && !error.empty());
  }
  CHECK(read_extension_package(root, "orphan", "", source, error) == OB_FILE_NOT_EXIST);
  CHECK(source.scripts_.empty());
  CHECK(read_extension_package(argv[2], "text_ops", "", source, error) == OB_SUCCESS);
  CHECK(source.native_module_.empty());
  CHECK(source.scripts_[0].sql_.find("CREATE FUNCTION seekdb_char_count") != std::string::npos);
  CHECK(read_extension_package(argv[2], "text_ops", "1.1", source, error) == OB_SUCCESS);
  CHECK(source.version_ == "1.1" && source.scripts_.size() == 2);
  CHECK(source.scripts_[1].sql_.find("CREATE FUNCTION seekdb_is_empty") != std::string::npos);
  CHECK(source.from_version_.empty());
  CHECK(read_extension_update(argv[2], "text_ops", "1.0", "1.1", source, error) == OB_SUCCESS);
  CHECK(source.from_version_ == "1.0" && source.version_ == "1.1" && source.scripts_.size() == 1);
  CHECK(source.scripts_[0].sql_.find("CREATE FUNCTION seekdb_char_count") == std::string::npos);
  CHECK(source.scripts_[0].sql_.find("CREATE FUNCTION seekdb_is_empty") != std::string::npos);
  CHECK(read_extension_update(root, "plain", "1.0", "2.0", source, error) == OB_FILE_NOT_EXIST);
  CHECK(source.from_version_.empty() && source.scripts_.empty() && !error.empty());
  CHECK(read_extension_update(root, "update_only", "new", "", source, error) == OB_SUCCESS);
  CHECK(source.from_version_ == "new" && source.version_ == "old" && source.scripts_.size() == 2);
  CHECK(source.scripts_[0].from_version_ == "new" && source.scripts_[0].to_version_ == "middle");
  CHECK(source.scripts_[1].from_version_ == "middle" && source.scripts_[1].to_version_ == "old");
  CHECK(source.scripts_[1].sql_ == "SELECT 'old';");
  std::ofstream(root + "/update_only/update_only--middle--old.sql") << "SELECT 'changed';";
  CHECK(source.scripts_[1].sql_ == "SELECT 'old';");
  for (const char *from : {"", "../x", "a--b"}) {
    CHECK(read_extension_update(root, "update_only", from, "old", source, error) == OB_INVALID_ARGUMENT);
    CHECK(source.name_.empty() && source.from_version_.empty() && source.scripts_.empty());
  }
  CHECK(read_extension_update(root, "noop", "v1", "", source, error) == OB_SUCCESS);
  CHECK(source.from_version_ == "v1" && source.version_ == "v1" && source.scripts_.empty());
  CHECK(read_extension_update(root, "noop_override", "v1", "", source, error) == OB_SUCCESS);
  CHECK(source.requires_ == std::vector<std::string>({"hidden"}) && source.prerequisites_.empty());
  CHECK(source.from_version_ == "v1" && source.scripts_.empty());
  CHECK(read_extension_update(root, "chain_override", "1.0", "", source, error) == OB_SUCCESS);
  CHECK(source.requires_ == std::vector<std::string>({"hidden_dependency"}) && source.prerequisites_.empty());
  for (const char *name : {"chain_escape", "chain_invalid", "update_total", "secondary_default",
                          "secondary_directory", "secondary_self", "secondary_large"}) {
    CHECK(read_extension_update(root, name, "1.0", "", source, error) == OB_INVALID_ARGUMENT);
    CHECK(source.name_.empty() && source.from_version_.empty() && source.scripts_.empty() && !error.empty());
  }
  CHECK(read_extension_update(root, "chain_empty", "1.0", "2", source, error) == OB_SUCCESS);
  CHECK(source.scripts_.size() == 1 && source.scripts_[0].sql_ == " \n");
  CHECK(read_extension_package(root, "chain_empty", "", source, error) == OB_SUCCESS);
  CHECK(source.from_version_.empty() && source.scripts_.size() == 2);
  CHECK(read_extension_update(root, "blank_update", "1.0", "", source, error) == OB_SUCCESS);
  CHECK(source.scripts_.size() == 1 && source.scripts_[0].sql_.empty());
  CHECK(source.from_version_ == "1.0" && source.version_ == "2");
  CHECK(read_extension_package(root, "versioned", "", source, error) == OB_SUCCESS);
  CHECK(source.requires_ == std::vector<std::string>({"gamma"}));
  CHECK(source.prerequisites_ == std::vector<std::string>({"alpha", "beta"}));
  CHECK(validate_extension_package_source(source, error) == OB_SUCCESS);
  const auto versioned_source = source;
  CHECK(read_extension_package(root, "dependency_limit", "", source, error) == OB_SUCCESS);
  CHECK(source.requires_.size() == 32 && source.prerequisites_.size() == 32);
  CHECK(validate_extension_package_source(source, error) == OB_SUCCESS);
  CHECK(read_extension_update(root, "dependency_overflow", "1.0", "2", source, error) == OB_SUCCESS);
  CHECK(source.requires_.size() == 33 && source.prerequisites_.empty());
  CHECK(versioned_source.prerequisites_ == std::vector<std::string>({"alpha", "beta"}));
  CHECK(read_extension_update(root, "versioned", "1.0", "tip", source, error) == OB_SUCCESS);
  CHECK(source.requires_ == std::vector<std::string>({"gamma"}));
  CHECK(source.prerequisites_ == std::vector<std::string>({"beta"}));
  CHECK(read_extension_update(root, "versioned", "middle", "tip", source, error) == OB_SUCCESS);
  CHECK(source.prerequisites_.empty());
  CHECK(read_extension_update(root, "versioned", "middle", "middle", source, error) == OB_SUCCESS);
  CHECK(source.requires_ == std::vector<std::string>({"beta"}) && source.prerequisites_.empty());
  std::ofstream(root + "/versioned/versioned--tip.control") << "requires = ''";
  CHECK(read_extension_package(root, "versioned", "", source, error) == OB_SUCCESS);
  CHECK(source.requires_.empty() && source.prerequisites_ == std::vector<std::string>({"alpha", "beta"}));
  std::ofstream(root + "/versioned/versioned--tip.control") << "# inherit primary, not middle";
  CHECK(read_extension_package(root, "versioned", "", source, error) == OB_SUCCESS);
  CHECK(source.requires_ == std::vector<std::string>({"base"}));
  CHECK(source.prerequisites_ == std::vector<std::string>({"alpha", "beta"}));
}

File: rust/plugin-runtime/tests/package_source.py
#!/usr/bin/env python3
"""Private filesystem fixtures for the real C++/Rust package source reader."""
import pathlib
import subprocess
import sys
import tempfile

with tempfile.TemporaryDirectory(prefix="seekdb-extension-package-") as temporary:
    root = pathlib.Path(temporary) / "packages"
    root.mkdir()

    # PG-style shared directory; exercise Rust preparation through the real
    # C ABI and C++ owned-source adapter, not only Rust unit tests.
    (root / "flat.control").write_text(
        "default_version = '2'\ndirectory = 'flat_sql'\n"
        "module_pathname = '$libdir/flat'\nnative_module = 'org.flat'\n", encoding="utf-8")
    flat_sql = root / "flat_sql"
    flat_sql.mkdir()
    (flat_sql / "flat--1.sql").write_text("SELECT 'MODULE_PATHNAME';", encoding="utf-8")
    (flat_sql / "flat--1--2.sql").write_text("SELECT 'MODULE_PATHNAME';", encoding="utf-8")
    (flat_sql / "flat--2.control").write_text("module_pathname = '$libdir/flat_v2'\n", encoding="utf-8")

    (root / "policy.control").write_text("default_version = '3'\nsuperuser = false\n", encoding="utf-8")
    (root / "policy--1.sql").write_text("SELECT 1;", encoding="utf-8")
    (root / "policy--1--2.sql").write_text("SELECT 2;", encoding="utf-8")
    (root / "policy--2--3.sql").write_text("SELECT 3;", encoding="utf-8")
    (root / "policy--2.control").write_text("superuser = true\n", encoding="utf-8")

    native = root / "native_only"
    native.mkdir()
    (native / "native_only.control").write_text("default_version = '1.0'\nnative_module = 'org.test'\ninstall_source = 'native'\n", encoding="utf-8")
    (native / "native_only--1.0--1.1.sql").write_text("SELECT 2;", encoding="utf-8")

    def package(name, control="default_version = '1.0'\n", sql=b"SELECT 1;\n"):
        directory = root / name
        directory.mkdir()
        (directory / f"{name}.control").write_text(control, encoding="utf-8")
        (directory / f"{name}--1.0.sql").write_bytes(sql)
        return directory

    plain = package("plain", "default_version = '1.0'\nrequires = 'dep1, dep2'\nrelocatable = true\n", b"SELECT 'first';\n")
    package("ambiguous")
    (root / "ambiguous.control").write_text("default_version = '1.0'\n", encoding="utf-8")
    (plain / "plain--2.0.sql").write_text("SELECT 'second';\n", encoding="utf-8")
    package("bad_utf8", sql=b"\xff")
    package("sql_nul", sql=b"SELECT 1;\x00SELECT 2;")
    package("empty_sql", sql=b" \n\t")
    package("control_utf8", "default_version = '1.0'\n").joinpath("control_utf8.control").write_bytes(b"\xff")
    package("control_large", "#" + "x" * (64 * 1024))
    package("exact_limit", sql=b"x" * (4 * 1024 * 1024))
    package("large", sql=b"x" * (4 * 1024 * 1024 + 1))
    package("bad_control", "default_version = '1.0'\ntrusted = true\n")
    package("selfdep", "default_version = '1.0'\nrequires = 'selfdep'\n")
    directory = package("directory")
    script = directory / "directory--1.0.sql"
    script.unlink()
    script.mkdir()
    no_default = package("no_default", "# caller must select a version\n")
    (no_default / "no_default--1.sql").write_text("SELECT 1;\n", encoding="utf-8")
    outside = pathlib.Path(temporary) / "outside"
    outside.mkdir()
    (outside / "escape.control").write_text("default_version = '1.0'\n", encoding="utf-8")
    (root / "escape").symlink_to(outside, target_is_directory=True)
    script_escape = package("script_escape") / "script_escape--1.0.sql"
    script_escape.unlink()
    (outside / "script.sql").write_text("SELECT 1;\n", encoding="utf-8")
    script_escape.symlink_to(outside / "script.sql")
    chain = package("chain", "default_version = 'tip'\n", b"SELECT 'base'; -- tail without newline")
    (chain / "chain--1.0--middle.sql").write_text("SELECT 'middle';", encoding="utf-8")
    (chain / "chain--middle--tip.sql").write_text("SELECT 'tip';", encoding="utf-8")
    # A direct installation wins over the longer chain; reading it must not
    # execute or try to read scripts from the longer path.
    (chain / "chain--direct.sql").write_text("SELECT 'direct';", encoding="utf-8")
    (chain / "chain--tip--direct.sql").write_text("SELECT 'unused';", encoding="utf-8")
    versioned = package("versioned", "default_version = 'tip'\nrequires = 'base'\n")
    (versioned / "versioned--1.0--middle.sql").write_text("SELECT 2;", encoding="utf-8")
    (versioned / "versioned--middle--tip.sql").write_text("SELECT 3;", encoding="utf-8")
    for version, dependency in (("1.0", "alpha"), ("middle", "beta"), ("tip", "gamma")):
        (versioned / f"versioned--{version}.control").write_text(f"requires = '{dependency}'\n", encoding="utf-8")
    # Runtime reads selected controls only; packaging validates all controls.
    (versioned / "versioned--unused.control").write_text("trusted = true\n", encoding="utf-8")
    for name, override in (
        ("secondary_default", "default_version = '2'"),
        ("secondary_directory", "directory = 'sql'"),
        ("secondary_self", "requires = 'secondary_self'"),
        ("secondary_schema", "schema = 'other'"),
        ("secondary_module", "native_module = 'org.other'"),
        ("secondary_large", "#" * (64 * 1024 + 1)),
    ):
        directory = package(name, "default_version = '2'\n")
        (directory / f"{name}--1.0--2.sql").write_text("SELECT 2;", encoding="utf-8")
        (directory / f"{name}--2.control").write_text(override, encoding="utf-8")
    directory = package("secondary_escape")
    (directory / "secondary_escape--1.0.control").symlink_to(outside / "escape.control")
    for name, count in (("dependency_limit", 32), ("dependency_overflow", 33)):
        directory = package(name, "default_version = '2'\n")
        (directory / f"{name}--1.0--2.sql").write_text("SELECT 2;", encoding="utf-8")
        for version, prefix in (("1.0", "initial"), ("2", "final")):
            names = ",".join(f"{prefix}_{i}" for i in range(count))
            (directory / f"{name}--{version}.control").write_text(f"requires = '{names}'\n", encoding="utf-8")
    for name in ("chain_escape", "chain_invalid", "chain_total", "chain_override", "chain_empty"):
        directory = package(name, "default_version = '2'\n")
        update = directory / f"{name}--1.0--2.sql"
        if name == "chain_escape":
            update.symlink_to(outside / "script.sql")
        elif name == "chain_invalid":
            update.write_bytes(b"\xff")
        elif name == "chain_total":
            update.write_bytes(b"x" * (4 * 1024 * 1024))
        elif name == "chain_empty":
            update.write_text(" \n", encoding="utf-8")
        else:
            update.write_text("SELECT 2;", encoding="utf-8")
            (directory / f"{name}--2.control").write_text("requires = 'hidden_dependency'\n", encoding="utf-8")
    orphan = package("orphan", "default_version = 'tip'\n")
    (orphan / "orphan--elsewhere--tip.sql").write_text("SELECT 2;", encoding="utf-8")
    update_only = package("update_only", "default_version = 'old'\n")
    (update_only / "update_only--1.0.sql").unlink()
    (update_only / "update_only--new--middle.sql").write_text("SELECT 'middle'; -- tail", encoding="utf-8")
    (update_only / "update_only--middle--old.sql").write_text("SELECT 'old';", encoding="utf-8")
    # Target installation bytes must not be read by an update path.
    (update_only / "update_only--old.sql").write_bytes(b"\xff")
    for name in ("noop", "noop_override"):
        directory = package(name, "default_version = 'v1'\n")
        (directory / f"{name}--1.0.sql").unlink()
        if name == "noop_override":
            (directory / f"{name}--v1.control").write_text("requires = 'hidden'\n", encoding="utf-8")
    blank = package("blank_update", "default_version = '2'\n")
    (blank / "blank_update--1.0--2.sql").write_bytes(b"")
    update_total = package("update_total", "default_version = '3'\n")
    (update_total / "update_total--1.0--2.sql").write_bytes(b"x" * (4 * 1024 * 1024))
    (update_total / "update_total--2--3.sql").write_bytes(b"x")
    subprocess.run([sys.argv[1], str(root), sys.argv[2]], check=True, timeout=20)
